### PR TITLE
cluster manager: implement CDS API

### PR DESCRIPTION
**Pull request by mattklein123**
Fixes https://github.com/lyft/envoy/issues/144

The only remaining piece that is missing is more graceful initialization
of CDS clusters on boot. (Basically we should only start listening once
we have done an initial init pass on any loaded CDS clusters). I will do
this in a follow up a this change is enough for folks to start testing basic
functionality.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @timperrett @lizan

File: docs/configuration/cluster_manager/cds.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+=========================
+
+The cluster discovery service (CDS) is an optional API that Envoy will call to dynamically fetch
+cluster manager members. Envoy will reconcile the API response and add, modify, or remove known
+clusters depending on what is required.
+
+.. code-block:: json
+
+  {
+    "cluster": "{...}",
```
Comment (timperrett) [inline docs/configuration/cluster_manager/cds.rst]: Small comment here: `"{...}"` might be a bit confusing for readers, as it looks like a JSON object inside a string, which would naturally not be accurate.

Comment (mattklein123) [inline docs/configuration/cluster_manager/cds.rst]: for better or worse this is the standard we have been using. I agree at this point just using {} would probably be better. I can do a bulk find/replace in a follow up change.

Comment (timperrett) [inline docs/configuration/cluster_manager/cds.rst]: No problem - its not an issue for me, was just thinking about the less familiar reader :-)

File: source/common/filter/auth/client_ssl.h [hunk 2]
```diff
@@ -88,7 +88,7 @@ class Config : public Http::RestApiFetcher {
   void createRequest(Http::Message& request) override;
   void parseResponse(const Http::Message& response) override;
   void onFetchComplete() override {}
-  void onFetchFailure(Http::AsyncClient::FailureReason reason) override;
+  void onFetchFailure(EnvoyException* e) override;
```
Comment (ccaraman) [inline source/common/filter/auth/client_ssl.h]: nit: full word for variable name.

**Pull request merged**